File: packages/core/src/schemas/metadata/MetadataProperty.ts
import {
  ClassProperties,
  IJsDocTagInfo,
  IMetadataSchema,
} from "@typia/interface";

import { IProtobufProperty } from "../protobuf/IProtobufProperty";
import { IMetadataDictionary } from "./IMetadataDictionary";
import { MetadataSchema } from "./MetadataSchema";

export class MetadataProperty {
  public readonly key: MetadataSchema;
  public readonly value: MetadataSchema;
  public readonly description: string | null;
  public readonly jsDocTags: IJsDocTagInfo[];
  public readonly mutability?: "readonly" | null | undefined;

  public of_protobuf_?: IProtobufProperty;

  /* -----------------------------------------------------------
    CONSTRUCTORS
  ----------------------------------------------------------- */
  private constructor(props: ClassProperties<MetadataProperty>) {
    this.key = props.key;
    this.value = props.value;
    this.description = props.description;
    this.jsDocTags = props.jsDocTags;
    this.mutability = props.mutability;
  }

  /** @internal */
  public static create(
    props: ClassProperties<MetadataProperty>,
  ): MetadataProperty {
    return new MetadataProperty(props);
  }

  /** @internal */
  public static from(
    property: IMetadataSchema.IProperty,
    dict: IMetadataDictionary,
  ) {
    return MetadataProperty.create({
      key: MetadataSchema.from(property.key, dict),
      value: MetadataSchema.from(property.value, dict),
      description: property.description,
      jsDocTags: property.jsDocTags.slice(),
      mutability: property.mutability,
    });
  }

  public toJSON(): IMetadataSchema.IProperty {
    return {
      key: this.key.toJSON(),
      value: this.value.toJSON(),
      description: this.description,
      jsDocTags: this.jsDocTags,
      mutability: this.mutability,
    };
  }
}

File: packages/core/src/schemas/metadata/MetadataProperty.go
package metadata

type MetadataProperty struct {
	Key         *MetadataSchema
	Value       *MetadataSchema
	Description *string
	JsDocTags   []JsDocTagInfo
	Mutability  *string

	OfProtobuf any
}

type MetadataPropertyProps struct {
	Key         *MetadataSchema
	Value       *MetadataSchema
	Description *string
	JsDocTags   []JsDocTagInfo
	Mutability  *string
	OfProtobuf  any
}

func CreateMetadataProperty(props MetadataPropertyProps) *MetadataProperty {
	return &MetadataProperty{
		Key:         props.Key,
		Value:       props.Value,
		Description: props.Description,
		JsDocTags:   cloneJsDocTags(props.JsDocTags),
		Mutability:  props.Mutability,
		OfProtobuf:  props.OfProtobuf,
	}
}

func MetadataPropertyFrom(property MetadataPropertyJSON, dict *IMetadataDictionary) (*MetadataProperty, error) {
	key, err := MetadataSchemaFrom(property.Key, dict)
	if err != nil {
		return nil, err
	}
	value, err := MetadataSchemaFrom(property.Value, dict)
	if err != nil {
		return nil, err
	}
	return CreateMetadataProperty(MetadataPropertyProps{
		Key:         key,
		Value:       value,
		Description: property.Description,
		JsDocTags:   property.JsDocTags,
		Mutability:  property.Mutability,
	}), nil
}

func (m *MetadataProperty) ToJSON() MetadataPropertyJSON {
	return MetadataPropertyJSON{
		Key:         m.Key.ToJSON(),
		Value:       m.Value.ToJSON(),
		Description: m.Description,
		JsDocTags:   cloneJsDocTags(m.JsDocTags),
		Mutability:  m.Mutability,
	}
}
